src/sys/futex.rs

```diff
@@ -0,0 +1,363 @@
+/// #   unistd::{fork,ForkResult},
+/// # };
+/// # use std::{
+/// #   time::{Instant, Duration},
+/// #   num::NonZeroUsize,
+/// #   mem::{ManuallyDrop, size_of},
+/// #   os::unix::io::OwnedFd,
+/// #   sync::Arc,
+/// #   thread::{spawn, sleep},
+/// # };
+/// const TIMEOUT: Duration = Duration::from_millis(500);
+/// const DELTA: Duration = Duration::from_millis(100);
+/// # fn main() -> nix::Result<()> {
+/// let futex: Futex = Futex::new(0);
+///
+/// // If the value of the futex is 0, wait for wake. Since the value is 0 and no wake occurs,
+/// // we expect the timeout will pass.
+///
+/// let instant = Instant::now();
+/// assert_eq!(futex.wait(0, Some(TIMEOUT)),Err(Errno::ETIMEDOUT));
+/// assert!(instant.elapsed() > TIMEOUT);
+///
+/// // If the value of the futex is 1, wait for wake. Since the value is 0, not 1, this will
+/// // return immediately.
+///
+/// let instant = Instant::now();
+/// assert_eq!(futex.wait(1, Some(TIMEOUT)),Err(Errno::EAGAIN));
+/// assert!(instant.elapsed() < DELTA);
+///
+/// // Test across threads
+/// // -------------------------------------------------------------------------
+///
+/// let futex = Arc::new(futex);
+/// let futex_clone = futex.clone();    
+/// let instant = Instant::now();
+/// spawn(move || {
+///     sleep(TIMEOUT);
+///     assert_eq!(futex_clone.wake(1),Ok(1));
+/// });
+/// assert_eq!(futex.wait(0, Some(2 * TIMEOUT)), Ok(()));
+/// assert!(instant.elapsed() > TIMEOUT && instant.elapsed() < TIMEOUT + DELTA);
+///
+/// // Test across processes
+/// // -------------------------------------------------------------------------
+///
+/// let shared_memory = unsafe { mmap::<OwnedFd>(
+///     None,
+///     NonZeroUsize::new_unchecked(size_of::<Futex<false>>()),
+///     ProtFlags::PROT_WRITE | ProtFlags::PROT_READ,
+///     MapFlags::MAP_SHARED | MapFlags::MAP_ANONYMOUS,
+///     None,
+///     0
+/// )? };
+/// let futex_ptr = shared_memory.cast::<Futex<false>>();
+/// let futex = unsafe { &*futex_ptr };
+/// match unsafe { fork()? } {
+///     ForkResult::Parent { child } => {
+///         sleep(TIMEOUT);
+///         assert_eq!(futex.wake(1),Ok(1));
+///         // Wait for child process to exit
+///         unsafe {
+///             assert_eq!(libc::waitpid(child.as_raw(), std::ptr::null_mut(), 0), child.as_raw());
+///         }
+///     },
+///     ForkResult::Child => {
+///         let now = Instant::now();
+///         assert_eq!(futex.wait(0, Some(2 * TIMEOUT)),Ok(()));
+///         assert!(now.elapsed() > TIMEOUT && now.elapsed() < TIMEOUT + DELTA);
+///     }
+/// }
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Debug)]
+pub struct Futex<const PRIVATE: bool = false>(UnsafeCell<u32>);
+impl<const PRIVATE: bool> Futex<PRIVATE> {
+    const MASK: i32 = if PRIVATE { libc::FUTEX_PRIVATE_FLAG } else { 0 };
+    /// Constructs new futex with a given `val`.
+    pub fn new(val: u32) -> Self {
+        Self(UnsafeCell::new(val))
+    }
+    /// If the value of the futex:
```

Comment:
The rest of the nix source puts newlines between structs, impls, functions, etc. I'd appreciate if we matched that style.

Reply:
Fixed.